serial-keel as a service

stephen-nordic opened the pull request.

stephen-nordic committed:
serial-keel as a service

stephen-nordic commented on serial-keel.template.service:
```diff
@@ -0,0 +1,13 @@
+RestartSec=1
+
+[Install]
+WantedBy=multi-user.target
```
So figured that specifying the `User` does indeed run the service as that user, but there doesn't seem to be any good way to use a variable to direct to the user home directory. What is being done here is that the script above fills in those details in the .service file before placing it in /etc/systemd/system/

torsteingrindvik commented on check_and_upgrade_serial_keel.sh:
```diff
@@ -0,0 +1,45 @@
+sudo systemctl stop serial-keel || true
+
+echo "STEP 4/5: Installing the new serial-keel build.."
+cargo install --bin serial-keel --path ./core
```
Missing `--features` here.

Blaming myself, will be addressed in #11 :slightly_smiling_face:

stephen-nordic replied:
ah, I didn't realize it needed to be mentioned in both build and install - will do a quick check and fix! So it should work with / without #11 :)

stephen-nordic replied:
Fixed this, and removed the unnecessary cargo build step from above.

torsteingrindvik commented:
I really like your approach :slightly_smiling_face: 

Could you add a section in the README about using this?
Especially I'm curious about the log output from the service.

Also maybe let's introduce a `/scripts` folder to the repo, and put this under `/scripts/systemd/<PR files>`.

stephen-nordic commented:
> I really like your approach slightly_smiling_face
> 
> Could you add a section in the README about using this? Especially I'm curious about the log output from the service.
> 
> Also maybe let's introduce a `/scripts` folder to the repo, and put this under `/scripts/systemd/<PR files>`.

Sure, I'll get those done :+1: 

The logs seem to be managed entirely by journalctl. For instance, `sudo journalctl -u serial-keel.service -e` shows the latest logs first. From a couple of google searches, the data appears to be in `/var/log/journal/*` in binary format, and I'm not sure they are parseable by anything other than journalctl. I will do some checking and let you know.

stephen-nordic committed:
Added README section for setting up serial-keel to run as a service, reorganized files, added mocks-share-endpoints flag to installation

stephen-nordic commented:
> I really like your approach slightly_smiling_face
> 
> Could you add a section in the README about using this? Especially I'm curious about the log output from the service.
> 
> Also maybe let's introduce a `/scripts` folder to the repo, and put this under `/scripts/systemd/<PR files>`.

Moved these files to a new scripts/systemd folder, and added a README section.

torsteingrindvik commented:
FYI right now there are duplicate files in the PR.

stephen-nordic committed:
Removed duplicate files

stephen-nordic commented:
> FYI right now there are duplicate files in the PR.

oops, thought I had moved them. Removed the duplicates from the repo root now.

torsteingrindvik commented on scripts/systemd/serial-keel.template.service:
```diff
@@ -0,0 +1,13 @@
+
+[Service]
+User=[user-name-here]
+WorkingDirectory=/home/[user-name-here]/serial-keel/
```
Should this matter?
Since we do `cargo install` the binary should be located in this user's PATH.

EDIT: So I don't see how a working dir should matter.

stephen-nordic replied:
Removed this WorkingDirectory line. However we needed to cd into the repo directory on service start to do the `git rev-parse`, which is now being done in the script itself.

torsteingrindvik commented on scripts/systemd/serial-keel.template.service:
```diff
@@ -0,0 +1,13 @@
+[Service]
+User=[user-name-here]
+WorkingDirectory=/home/[user-name-here]/serial-keel/
+ExecStartPre=/home/[user-name-here]/serial-keel/scripts/systemd/check_and_upgrade_serial_keel.sh [branch-here]
```
Since we do `git rev-parse --show-toplevel`, we have the path to the repo.
Could we put that path here instead of `/home/[user-name-here]`?
Then the user can have the `serial-keel` folder anywhere.

stephen-nordic replied:
Changed the script so that we could have the serial keel folder anywhere

torsteingrindvik commented on scripts/systemd/serial-keel.template.service:
```diff
@@ -0,0 +1,13 @@
+User=[user-name-here]
+WorkingDirectory=/home/[user-name-here]/serial-keel/
+ExecStartPre=/home/[user-name-here]/serial-keel/scripts/systemd/check_and_upgrade_serial_keel.sh [branch-here]
+ExecStart=/home/[user-name-here]/.cargo/bin/serial-keel /home/[user-name-here]/config.ron
```
So since this is executing as the user, I'm hoping the `env` is inherited.
Then `/home/$user/.cargo/bin` should be on the `PATH`.

In that case we _should_ be able to just `ExecStart=serial-keel ...`.

If that is not possible (we can't inherit the users' env) then I propose we append `/home/[user-name-here]/.cargo/bin` to the `PATH` manually.

I found this: https://wiki.archlinux.org/title/systemd/User
Which shows use of `~/.config/environment.d/`

torsteingrindvik commented on scripts/systemd/check_and_upgrade_serial_keel.sh:
```diff
@@ -0,0 +1,41 @@
+# This replaces all occurences of [user-name-here] and [branch-here] in the serial-keel.template file and 
+# with the current user and branch name and places the service file in /etc/systemd/system
+CURRENT_USER=$(whoami)
+sudo sh -c 'sed "s/\[user-name-here\]/'"$CURRENT_USER"'/g;
```
I saw here: https://wiki.archlinux.org/title/systemd/User
That `~/.config/systemd/user/` is possible. That should eliminate all needs for `sudo`.
Then we `systemctl --user` everywhere instead.

stephen-nordic replied:
Now using `systemctl --user`, and removed all references to `sudo`.

torsteingrindvik commented on README.md:
```diff
@@ -541,3 +542,31 @@ There is a WIP GUI which can be started via (from the root folder) `cargo r --bi
+You can setup Serial Keel to run as a systemd service on startup. In `scripts/systemd`, you'll find a template `.service`
+file and a setup script.
+
+The script assumes that you have a valid serial-keel configuration file in your user home directory named `config.ron`,
```
It would be great if the script would use `~/config.ron` as the default but allow passing it in as well.

stephen-nordic replied:
Done :+1:

torsteingrindvik replied:
Nice! Docs should be updated to match :slightly_smiling_face:

stephen-nordic committed:
Addressed most review comments

torsteingrindvik commented on scripts/systemd/check_and_upgrade_serial_keel.sh:
```diff
@@ -0,0 +1,68 @@
+# If not, we have a serial-keel update
+echo "We have a serial-keel update to perform.."
+mkdir -p $USER_HOME/.config/environment.d/
+env > $USER_HOME/.config/environment.d/env
```
Perhaps surround the making of this folder and file with a test if it exists?
Since `env` is only populated when the user runs this script manually, and thereafter it would only be overwritten with the same contents by the service.

torsteingrindvik commented on scripts/systemd/check_and_upgrade_serial_keel.sh:
```diff
@@ -0,0 +1,68 @@
+# If not, we have a serial-keel update
+echo "We have a serial-keel update to perform.."
+mkdir -p $USER_HOME/.config/environment.d/
+env > $USER_HOME/.config/environment.d/env
```
Also I think the file _must_ have a `.conf` ending, so either `env.conf` or `serial-keel.conf` or similar.

Mangl10 replied:
agreed. Probably `serial-keel.conf` is clearer in intent.

torsteingrindvik commented on scripts/systemd/check_and_upgrade_serial_keel.sh:
```diff
@@ -0,0 +1,68 @@
+
+# The name of the branch can be specified as a argument, or it defaults to 'main'
+BRANCH=${1-main}
+CONFIG_PATH=${2-/home/$CURRENT_USER/config.ron}
```
You can reuse the home var here.

stephen-nordic replied:
Added a commit that resolves the remaining comments.

stephen-nordic committed:
Addressed review comments and updated README

torsteingrindvik committed:
Update SHA to see how reboot is handled

Signed-off-by: Torstein Grindvik <torstein.grindvik@nordicsemi.no>

torsteingrindvik commented on scripts/systemd/check_and_upgrade_serial_keel.sh:
```diff
@@ -0,0 +1,67 @@
+systemctl --user daemon-reload
+
+echo "STEP 2/4: Stopping serial-keel server if already running.."
+systemctl --user stop serial-keel || true
```
So I did some testing and I think we have to split things up a bit.

1. A single "one time script" to do the copying of the template service file and enabling the service
2. Everything else: Check git, upgrade if needed, start the server

Because I think that doing e.g. `systemctl --user stop serial-keel` as part of the `serial-keel` systemd service itself is a strange "self reference" which kills its own upgrade.

**How to see this behaviour:**
First I ran the script to install the service.

Notice I pushed an empty commit to your branch.
This allows me locally then `git reset --hard` to the commit _before_ the empty commit.
If I then do `systemctl --user restart serial-keel` it then triggers the update because it sees the upstream has a never revision (since `reset --hard` removes the local one).

It will then run successfully until after
`echo "STEP 2/4: Stopping serial-keel server if already running.."`
Then the update process gets killed, and the status (`systemctl --user status serial-keel`) says:

```text
Feb 25 11:36:19 pop-os check_and_upgrade_serial_keel.sh[6474]: STEP 2/4: Stopping serial-keel server if already running..
Feb 25 11:36:19 pop-os systemd[2677]: serial-keel.service: Control process exited, code=killed, status=15/TERM
Feb 25 11:36:19 pop-os systemd[2677]: serial-keel.service: Failed with result 'signal'.
Feb 25 11:36:19 pop-os systemd[2677]: Stopped Serial Keel Server.
```

Which leads me to think `systemctl --user stop serial-keel || true` sent a terminate signal to _itself_.

Therefore I think we need a cleaner separation where the `serial-keel` script has no references to `systemctl`.

Similarly, it shouldn't make sense to finish off the script with `systemctl --user start serial-keel`, since if we are assuming that it's `systemd` that is running this script then it should already be in the context of starting (or restarting) the service.

Let me know if you have any questions.

Mangl10 replied:
wouldn't it be better to skip step number 2 and on step 4, instead just do `systemctl --user restart serial-keel`? I assume that would achieve the same result and you can still keep it in one script?

stephen-nordic replied:
I suppose then, we will have an issue with overwriting the serial-keel binary when it is running. I think I'll split the script into two, one which is only triggered by the user - and the other run as a pre-execution script by the service.

torsteingrindvik replied:
@Mangl10 I think there shouldn't be any calls to `systemctl` except the first time when the service is installed.
When the script is called after the first time we assume it is being called by `systemd`, which handles starting/restarting by itself, so we don't have to.

Therefore I think a valid strategy is to either:

1. Keep everything in one script, but have a guard which checks if this is the first time we are running the script (e.g. by looking for the service file), and only do `systemctl` calls within this guard, or
2. Split it up into two scripts, one for service setup, and one for running the upgrade+server itself.

@stephen-nordic I don't that will be an issue.
Remember, when `systemd` is (re)starting the service it will stop it first (if it was active) which would first stop the server.
That is not the same as starting the service and then running the script locally- that isn't a use case we need or want to support.

stephen-nordic replied:
@torsteingrindvik Ah yes, I see that makes sense. So @Mangl10's suggestion would work.

But if we were to split the script into two, (or within the guard for first-time script run in the single script method) I don't think we would be able to eliminate all calls to systemctl in the upgrade script. What if there was a change in the .service file itself? In that case, we would need to update the .service file, do a systemctl daemon-reload and restart the service, both of which would involve calls to systemctl.

Or do we not want to support automatic .service file updates entirely? We could make it such that .service file changes can only be carried out by manually.

I'd like to propose this:
 1. Keeping everything in one script, and have a check to see if we are running the script for the first time. We will have systemctl calls only within this check. If the .service file has changed, the user has to run this script manually passing a special flag, which will run the script as if it were a first time installation, stop the serial-keel service and replace the service file.
 
 Edit:
 (or)
 2. We could tolerate calls to systemctl --daemon-reload / restart in the upgrade section, unlike systemctl stop which would result in self-termination?

torsteingrindvik replied:
Right now I'm thinking it's fine to just update the service manually, that shouldn't happen often.

So in script 1 just copy the template and overwrite if the service exists etc.
And in script 2 perhaps call script 1 if first time, and else skip that part.
(EDIT: This is my main rationale for having two separate scripts, the user can then test either in isolation, and e.g. just call the first one to upgrade/install the service, and try calling the second one to test upgrades of SK).

I don't think automating the service upgrade itself is worth it for now, our main goal is just to make sure the SK server is up to date.

stephen-nordic replied:
Thanks @torsteingrindvik and @Mangl10 :slightly_smiling_face: I've added a commit that should work. Let me know  if there are any more concerns. :+1:

stephen-nordic committed:
Split script into two - one purely for install (triggered manually) and the other that checks for serial-keel upgrades on reboot

stephen-nordic committed:
If we happened to call the upgrade script first, call the install script instead

torsteingrindvik commented on README.md:
````diff
@@ -541,3 +542,28 @@ There is a WIP GUI which can be started via (from the root folder) `cargo r --bi
+- Once done, the serial-keel service should be running. Running `systemctl --user status serial-keel.service`
+should let you know whether the service started successfully.
+
+- You can access the latest logs of the serial-keel service with:
+```
+journalctl --user-unit serial-keel.service -e
+```
````
```suggestion
- Logs are accessed via `journalctl`. Check `man journalctl` for usage. Here is an example:
```
journalctl --user-unit serial-keel --pager-end --since "2 days ago" --output=cat
```
The `--output=cat` is a "raw" logging mode, this ensures we don't get timestamps twice.
```

torsteingrindvik commented:
Great job! Tried it locally and the service seems to work well, including upgrades.
